p2p/router.go

```diff
@@ -585,55 +587,83 @@ func (r *Router) openConnection(ctx context.Context, conn Connection) {
 func (r *Router) dialPeers() {
 	r.logger.Debug("starting dial routine")
 	ctx := r.stopCtx()
+
+	addresses := make(chan NodeAddress)
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < runtime.NumCPU(); i++ {
```

Comment:
So if I'm understanding correctly, this loop spawns off `runtime.NumCPU()` number of goroutines, where each goroutine handles reading off a channel and attempts to dial, correct? What's the motivation behind this change and why this specific number of loops? How does this help and why does it seem to work better?


Not questing that this is not a correct approach, I just want to really make sure we document this well and understand why this seems to work better.

Would love to even get thoughts from @odeke-em and maybe even @erikgrinaker (if he has time) 🙏

Reply:
The number of threads that we use is totally arbitrary, and I suspect fewer is generally better, but too few might be a problem, so picking the number isn't perfect. 

From a high level, the dialing is in a "single-producer, multi-consumer/worker" pattern meant to provide some parallelism, but to also provide a throttle and avoid unbounded parallelism for dialing (and therefor potentially spam the network, or use unbounded local resources.

Reply:
Thanks for the tag @alexanderbez! So essentially @tychoish using runtime.NumCPU() allows the Go runtime to multiplex each of those goroutines on a different CPU for concurrency, but also when making network connections and in IO they'll be on independent OS threads too. I think this is a good approach and it scales well.